### PR TITLE
Added SlimefunGuideOpenEvent.

### PR DESCRIPTION
**Pull request by LinoxGH**
## Description

Added a new event to the api which is fired when SlimefunGuide is opened.
Also did some refactoring.

## Changes

New event!
Some refactoring in SlimefunGuideListener.

## Checklist

- [ ] I have fully tested the proposed changes and promise that they will not break everything into chaos.
- [ ] I have also tested the proposed changes in combination with various popular addons and can confirm my changes do not break them.
- [X] I followed the existing code standards and didn't mess up the formatting.
- [X] I did my best to add documentation to any public classes or methods I added.
- [X] I have added `Nonnull` and `Nullable` annotations to my methods to indicate their behaviour for null values
- [ ] I added sufficient Unit Tests to cover my code.

### CODE REVIEW

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/SlimefunGuideListener.java [hunk 1]
```diff
@@ -72,16 +74,24 @@ else if (openGuide(e, SlimefunGuideLayout.CHEAT_SHEET) == Result.ALLOW) {
...
+            e.cancel();
+            SlimefunGuide.openGuide(p, layout);
+        } 
+    } 
 
     @Nonnull
     @ParametersAreNonnullByDefault
-    private Result openGuide(PlayerRightClickEvent e, SlimefunGuideLayout layout) {
-        Player p = e.getPlayer();
+    private Result tryGuide(Player p, PlayerRightClickEvent e, SlimefunGuideLayout layout) {
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/SlimefunGuideListener.java:91]: `tryOpenGuide` sounds a bit better in my opinion.

File: src/main/java/io/github/thebusybiscuit/slimefun4/api/events/GuideOpenEvent.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+import org.bukkit.event.HandlerList;
+import org.bukkit.inventory.ItemStack;
+
+/**
+ * This {@link Event} is called whenever a {@link Player} tries to open the Slimefun Guide book. 
+ * 
+ * @author Linox
+ *
+ */
+public class GuideOpenEvent extends Event implements Cancellable {
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/api/events/GuideOpenEvent.java:18]: * SlimefunGuideOpenEvent maybe?

File: src/main/java/io/github/thebusybiscuit/slimefun4/api/events/GuideOpenEvent.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+    @Nonnull
+    public Player getPlayer() {
+        return player;
+    }
+
+    @Nonnull
+    public ItemStack getGuide() {
+        return guide;
+    }
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/api/events/GuideOpenEvent.java:41]: Some docs comments for these two methods would be nice.

File: src/main/java/io/github/thebusybiscuit/slimefun4/api/events/GuideOpenEvent.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+ */
+public class GuideOpenEvent extends Event implements Cancellable {
+
+    private static final HandlerList handlers = new HandlerList();
+
+    private final Player player;
+    private final ItemStack guide;
+    private boolean cancelled;
+
+    public GuideOpenEvent(@Nonnull Player p, @Nonnull ItemStack guide) {
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/api/events/GuideOpenEvent.java:26]: I think a `SlimefunGuideLayout` with respective getters/setters would be great too

**Review by TheBusyBiscuit — CHANGES_REQUESTED** [review]

File: src/main/java/io/github/thebusybiscuit/slimefun4/api/events/SlimefunGuideOpenEvent.java [hunk 5]
```diff
@@ -9,36 +9,59 @@
...
+     * tries to open the Slimefun Guide with.
+     */
     @Nonnull
     public ItemStack getGuide() {
         return guide;
     }
+    
+    /**
+     * This returns the {@link SlimefunGuideLayout} of the Slimefun Guide
+     * that {@link Player} tries to open.
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/api/events/SlimefunGuideOpenEvent.java:59]: Missing `@return` tag

File: src/main/java/io/github/thebusybiscuit/slimefun4/api/events/SlimefunGuideOpenEvent.java [hunk 6]
```diff
@@ -9,36 +9,59 @@
...
+     * the Slimefun Guide.
+     */
     @Nonnull
     public Player getPlayer() {
         return player;
     }
 
+    /**
+     * This returns the {@link ItemStack} that {@link Player} 
+     * tries to open the Slimefun Guide with.
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/api/events/SlimefunGuideOpenEvent.java:50]: Missing `@return` tag

File: src/main/java/io/github/thebusybiscuit/slimefun4/api/events/SlimefunGuideOpenEvent.java [hunk 7]
```diff
@@ -9,36 +9,59 @@
...
         Validate.notNull(guide, "Guide cannot be null");
+        Validate.notNull(layout, "Layout cannot be null");
         this.player = p;
         this.guide = guide;
+        this.layout = layout;
     }
 
+    /**
+     * This returns the {@link Player} that tries to open 
+     * the Slimefun Guide.
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/api/events/SlimefunGuideOpenEvent.java:41]: Missing `@return` tag

**Review by TheBusyBiscuit — CHANGES_REQUESTED** [review]

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/SlimefunGuideListener.java [hunk 8]
```diff
@@ -72,16 +74,24 @@ else if (openGuide(e, SlimefunGuideLayout.CHEAT_SHEET) == Result.ALLOW) {
             }
         }
     }
+    
+    @ParametersAreNonnullByDefault
+    private void openGuide(Player p, PlayerRightClickEvent e, SlimefunGuideLayout layout) {
+        GuideOpenEvent event = new GuideOpenEvent(p, e.getItem(), layout) ;
```
Comment (poma123) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/SlimefunGuideListener.java:80]: Haha, build is failing.
```suggestion
        SlimefunGuideOpenEvent event = new SlimefunGuideOpenEvent(p, e.getItem(), layout);
```

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/SlimefunGuideListener.java [hunk 9]
```diff
@@ -11,6 +12,7 @@
 import org.bukkit.event.player.PlayerJoinEvent;
 import org.bukkit.inventory.ItemStack;
 
+import io.github.thebusybiscuit.slimefun4.api.events.GuideOpenEvent;
```
Comment (poma123) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/SlimefunGuideListener.java:15]:
```suggestion
import io.github.thebusybiscuit.slimefun4.api.events.SlimefunGuideOpenEvent;
```

**Review by poma123 — CHANGES_REQUESTED** [review]

**Review by TheBusyBiscuit — APPROVED** [review]
I would probably like a setter for the layout but this is good as it is

**Review by poma123 — APPROVED** [review]
LGTM

**Comment by LinoxGH** [comment]
There ya go

**Review by poma123 — DISMISSED** [review]

File: src/main/java/io/github/thebusybiscuit/slimefun4/api/events/SlimefunGuideOpenEvent.java [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+    }
+    
+    /**
+     * Changes the {@link SlimefunGuideLayout} that was tried to be opened with.
+     *
+     * @param layout
+     *                   The new {@link SlimefunGuideLayout}
+     */
+    public void setGuideLayout(@Nonnull SlimefunGuideLayout layout) {
+        this.layout = layout;
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/api/events/SlimefunGuideOpenEvent.java:79]:
```suggestion
        Validate.notNull(layout, "You must specify a layout that is not-null!");
        this.layout = layout;

```

**Review by poma123 — APPROVED** [review]

**Review by TheBusyBiscuit — APPROVED** [review]

**Comment by TheBusyBiscuit** [comment]
I only just realized that this might also be pretty easy to unit-test too, if none of you wants to do it, I'll push some unit tests for this event soon XD
